Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -63,8 +77,27 @@ void VirtIOConsole::handle_queue_update(u16 queue_index)
     VERIFY(queue_index <= TRANSMITQ);
     switch (queue_index) {
     case RECEIVEQ: {
-        ScopedSpinLock lock(get_queue(RECEIVEQ).lock());
-        get_queue(RECEIVEQ).discard_used_buffers(); // TODO: do something with incoming data (users writing into qemu console) instead of just clearing
+        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
+        auto& queue = get_queue(RECEIVEQ);
+        ScopedSpinLock queue_lock(queue.lock());
+        size_t used;
+        VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
+
+        auto used_space = m_receive_buffer->reserve_space(used).value();
+        auto remaining_space = RINGBUFFER_SIZE - used;
+
+        // Our algorithm always has only one buffer in the queue.
+        size_t unused;
+        VERIFY(queue.pop_used_buffer_chain(unused).is_empty());
```

Comment:
VERIFY statements with side-effects are weird, how about just:
```suggestion
        VERIFY(!queue.new_data_available());
```